Add NICE method

Jamie001129 opened the pull request:
**Implementation Details**
(1) 4 variants implemented: none, sparsity, proximity, plausibility
(2) Source: Adapted from official NICE repository (https://github.com/DBrughmans/NICE)
(3) Paper: Brughmans et al. (2024) "NICE: an algorithm for nearest instance counterfactual explanations" Data Mining and Knowledge Discovery
(4) Dataset: Adult
(5) Predictive model: Random Forest and MLP

**Potential Differences**
(1) Original autoencoder's structure was not provided so we have to create our own
(2) The 200 test samples were originally chosen at random so our samples could be different
(3) Run time will be different but the rankings of the four variants are the same

**Reproduced Results**
(1) RF as predictive model (updated on 11/13/2025)
| Variant      | Coverage  | CPU (ms) | Sparsity        | Proximity (L1)  | Plausibility    |
|--------------|-----------|----------|-----------------|-----------------|-----------------|
| none         | 200/200   | 20.11    | 3.19 ± 1.07     | 0.52 ± 0.52     | 0.0885 ± 0.0209 |
| sparsity     | 200/200   | 53.58    | 1.61 ± 0.91     | 0.34 ± 0.42     | 0.0886 ± 0.0220 |
| proximity    | 200/200   | 56.70    | 1.77 ± 1.01     | 0.32 ± 0.41     | 0.0887 ± 0.0224 |
| plausibility | 200/200   | 70.33    | 2.15 ± 1.08     | 0.38 ± 0.43     | 0.0892 ± 0.0227 |

(2) MLP as predictive model
| Variant | Coverage | CPU (ms) | Sparsity | Proximity (HEOM) | Plausibility (AE error) |
|---------|----------|----------|----------|------------------|-------------------------|
| none | 200/200 | 6.70 | 3.89 ± 1.37 | 3.14 ± 1.34 | 0.2044 ± 0.0319 |
| sparsity | 200/200 | 10.80 | 1.22 ± 0.50 | 1.06 ± 0.52 | 0.2008 ± 0.0360 |
| proximity | 200/200 | 11.34 | 1.41 ± 0.69 | 1.09 ± 0.67 | 0.2034 ± 0.0346 |
| plausibility | 200/200 | 19.68 | 2.37 ± 1.37 | 2.04 ± 1.24 | 0.2014 ± 0.0342 |

**Files Added/Modified**
Main implementation:
methods/catalog/nice/model.py - Main NICE wrapper class implementing RecourseMethod interface
methods/catalog/nice/reproduce.py - Comprehensive test reproducing paper results (part of table 6)

Library components:
methods/catalog/nice/library/__init__.py - Library exports
methods/catalog/nice/library/autoencoder.py - Autoencoder for plausibility measurement
methods/catalog/nice/library/data.py - Data handling and candidate filtering
methods/catalog/nice/library/distance.py - HEOM distance metric implementation
methods/catalog/nice/library/heuristic.py - Best-first greedy search
methods/catalog/nice/library/reward.py - Three reward functions (sparsity, proximity, plausibility)

Integration:
Updated methods/__init__.py to export NICE
Updated methods/catalog/__init__.py to include NICE

Jamie001129 committed:
complete nice implementation and update inits

Jamie001129 committed:
complete nice implementation and update inits

Commit:
creating stuff in the reproduce.py and commiting from the fxxking public pc

Commit:
creating stuff in the reproduce.py and commiting from the fxxking public pc

Commit:
fix the autoencoder and make a more comprehensive reproduce.py

Commit:
fix the autoencoder and make a more comprehensive reproduce.py

Commit:
add testing with mlp model to reproduce.py

Commit:
add testing with mlp model to reproduce.py

Commit:
modify reproduce to use the correct/author's way (HEOM distance)
to calculate proximity

Commit:
modify reproduce to use the correct/author's way (HEOM distance)
to calculate proximity

zkhotanlou commented on methods/catalog/nice/reproduce.py:
```diff
@@ -0,0 +1,457 @@
+    std_ae_error = ae_errors.std()
+    
+    # ============================================
+    # PRINT ALL FOUR METRICS (like Table 5 in paper)
```
I ran these tests locally, but the printed results don’t match the values reported in the paper. Please turn these print statements into assertions using the numbers from the table (a small tolerance is acceptable).

zkhotanlou commented on methods/catalog/nice/reproduce.py:
```diff
@@ -0,0 +1,457 @@
+    elif optimization == "none":
+        # None should be very plausible (it's an actual instance!)
+        # But we allow some tolerance since we measure on test set
+        assert avg_ae_error <= 0.02, \
```
I couldn’t find where the paper reports the average error rate. The only place I see something similar is in Table 7, but that value seems different from what’s being checked here. Could you point me to the exact reference?

Commit:
Change the autoencoder according to the atructure in repo NICE_experiments

Commit:
Change the autoencoder according to the atructure in repo NICE_experiments

Commit:
fix bugs and change print to assertion in reproduce.py

Commit:
fix bugs and change print to assertion in reproduce.py

Jamie001129 commented on methods/catalog/nice/reproduce.py:
```diff
@@ -0,0 +1,457 @@
+    for opt in ["none", "sparsity", "proximity", "plausibility"]:
+        nice = NICE(mlmodel=model, hyperparams={"optimization": opt})
+        
+        # Measure CPU time
```
I have removed cpu time assertions

zkhotanlou reviewed (commented):
Also, please fetch the changes from the main branch so that the pre-commit hooks could be run successfully.

Jamie001129 committed:
Merge branch 'charmlab:main' into feature/add-nice

Commit:
set seed

Commit:
set seed

Commit:
remove seed settings

Commit:
remove seed settings

Commit:
Start to fix the NICE implementation:
use the author's data and model

Commit:
Start to fix the NICE implementation:
use the author's data and model

Commit:
Merge upstream/main into feature/add-nice

Commit:
Merge upstream/main into feature/add-nice

Commit:
Merge branch 'feature/add-nice' of https://github.com/Jamie001129/recourse_benchmarks into feature/add-nice

Commit:
put back the
    NICE,
    Probe,
in methods/__init__.py

Commit:
model tuning

Commit:
Add OHE_minmax preprocessor matching NICE_experiments (preprocessing.py)
Raw data: label-encoded categorical + original scale continuous
Preprocessor: One-Hot Encoding + MinMaxScaler(-1, 1)

Fix autoencoder training on preprocessed data

Commit:
update test auc

Commit:
add mlp tuning

Jamie001129 committed:
Merge branch 'charmlab:main' into feature/add-nice

Jamie001129 committed:
Merge branch 'charmlab:main' into feature/add-nice
